File: server/text-processor.ts
import type { ProcessingConfig } from "@shared/schema";
import { chunkTextPreservingStructure } from "@shared/text-utils";
import { llmService } from "./llm-service";

export interface ProcessingProgress {
  chunkIndex: number;
  processedText: string;
  status: "success" | "retry" | "failed";
  retryCount: number;
  lastChunkMs?: number;
  avgChunkMs?: number;
  etaMs?: number;
  // Token/cost metrics for this chunk (optional)
  inputTokens?: number;
  outputTokens?: number;
  inputCost?: number;
  outputCost?: number;
  totalInputTokens?: number;
  totalOutputTokens?: number;
  totalCost?: number;
}

export interface TextProcessingResult {
  text: string;
  totalChunks: number;
  failedChunkIndexes: number[];
  totalInputTokens: number;
  totalOutputTokens: number;
  totalCost: number;
}

export class ProcessingCancelledError extends Error {
  constructor() {
    super("Processing cancelled");
    this.name = "ProcessingCancelledError";
  }
}

function waitForRetry(signal?: AbortSignal): Promise<void> {
  if (signal?.aborted) return Promise.reject(new ProcessingCancelledError());
  return new Promise((resolve, reject) => {
    const onAbort = () => {
      clearTimeout(timer);
      reject(new ProcessingCancelledError());
    };
    const timer = setTimeout(() => {
      signal?.removeEventListener("abort", onAbort);
      resolve();
    }, 1000);
    signal?.addEventListener("abort", onAbort, { once: true });
  });
}

export class TextProcessor {
  async processText(
    text: string,
    config: ProcessingConfig,
    onProgress: (progress: ProcessingProgress) => void,
    shouldCancel: () => boolean = () => false,
    signal?: AbortSignal
  ): Promise<TextProcessingResult> {
    const chunks = chunkTextPreservingStructure(text, config.batchSize);
    const processedChunks: string[] = [];
    const totalChunks = chunks.length;
    let durationsTotal = 0;
    let processedCount = 0;
    let totalInTokens = 0, totalOutTokens = 0, totalCost = 0;
    const failedChunkIndexes: number[] = [];

    for (let i = 0; i < chunks.length; i++) {
      if (shouldCancel()) throw new ProcessingCancelledError();
      const chunk = chunks[i].text;
      let retryCount = 0;
      let success = false;
      let processedText = "";
      const chunkStart = Date.now();

      // Try processing with one retry on failure
      while (retryCount < 2 && !success) {
        if (shouldCancel()) throw new ProcessingCancelledError();
        try {
          const result = await llmService.processChunk({
            text: chunk,
            cleaningOptions: config.cleaningOptions,
            speakerConfig: config.speakerConfig,
            modelSource: config.modelSource,
            modelName: config.modelName,
            ollamaModelName: (config as any).ollamaModelName,
            temperature: (config as any).temperature,
            llmCleaningDisabled: (config as any).llmCleaningDisabled === true,
            customInstructions: config.customInstructions,
            singlePass: (config as any).singlePass === true,
            extendedExamples: (config as any).extendedExamples === true,
            signal,
          });
          processedText = result.text;
          const chunkIn = result.usage.inputTokens || 0;
          const chunkOut = result.usage.outputTokens || 0;
          const chunkCost = (result.usage.inputCost || 0) + (result.usage.outputCost || 0);
          totalInTokens += chunkIn; totalOutTokens += chunkOut; totalCost += chunkCost;

          // Validate output
          const validation = await llmService.validateOutput(
            chunk,
            processedText,
            config
          );

          if (validation.valid) {
            success = true;
            processedChunks.push(processedText);
            const lastChunkMs = Date.now() - chunkStart;
            durationsTotal += lastChunkMs;
            processedCount += 1;
            const avgChunkMs = durationsTotal / processedCount;
            const remaining = totalChunks - (i + 1);
            const etaMs = Math.max(0, Math.round(avgChunkMs * remaining));
            
            onProgress({
              chunkIndex: i,
              processedText,
              status: "success",
              retryCount,
              lastChunkMs,
              avgChunkMs,
              etaMs,
              inputTokens: chunkIn,
              outputTokens: chunkOut,
              inputCost: result.usage.inputCost,
              outputCost: result.usage.outputCost,
              totalInputTokens: totalInTokens,
              totalOutputTokens: totalOutTokens,
              totalCost,
            });
          } else {
            retryCount++;
            if (retryCount < 2) {
              onProgress({
                chunkIndex: i,
                processedText: "",
                status: "retry",
                retryCount,
              });
              
              // Wait a bit before retry
              await waitForRetry(signal);
              if (shouldCancel()) throw new ProcessingCancelledError();
            }
          }
        } catch (error) {
          if (
            error instanceof ProcessingCancelledError ||
            signal?.aborted ||
            (error instanceof Error && error.name === "AbortError")
          ) {
            throw new ProcessingCancelledError();
          }
          retryCount++;
          if (retryCount < 2) {
            onProgress({
              chunkIndex: i,
              processedText: "",
              status: "retry",
              retryCount,
            });
            
            await waitForRetry(signal);
            if (shouldCancel()) throw new ProcessingCancelledError();
          }
        }
      }

      // If still failed after retries, use original chunk
      if (!success) {
        failedChunkIndexes.push(i);
        processedChunks.push(chunk);
        const lastChunkMs = Date.now() - chunkStart;
        durationsTotal += lastChunkMs;
        processedCount += 1;
        const avgChunkMs = durationsTotal / processedCount;
        const remaining = totalChunks - (i + 1);
        const etaMs = Math.max(0, Math.round(avgChunkMs * remaining));
        onProgress({
          chunkIndex: i,
          processedText: chunk,
          status: "failed",
          retryCount,
          lastChunkMs,
          avgChunkMs,
          etaMs,
        });
      }
    }

    return {
      text: processedChunks
        .map((processed, index) => `${processed}${chunks[index]?.separatorAfter ?? ""}`)
        .join(""),
      totalChunks,
      failedChunkIndexes,
      totalInputTokens: totalInTokens,
      totalOutputTokens: totalOutTokens,
      totalCost,
    };
  }
}

export const textProcessor = new TextProcessor();

File: gradio_app/text_processor.py
from __future__ import annotations

import re
import time
from typing import Callable, List, Optional

from .llm_service import LLMService, ProcessOptions, llm_service
from .models import (
    CleaningOptions,
    ProcessingConfig,
    ProcessingProgress,
    ProcessingSummary,
)
from .text_cleaner import apply_deterministic_cleaning


class TextProcessor:
    def __init__(self, service: LLMService | None = None) -> None:
        self.service = service or llm_service

    def split_into_chunks(self, text: str, batch_size: int) -> List[str]:
        sentences = re.findall(r"[^.!?]+(?:[.!?]+|$)", text) or [text]
        sentences = [s.strip() for s in sentences if s.strip()]
        chunks: List[str] = []
        current: List[str] = []
        for sentence in sentences:
            current.append(sentence)
            if len(current) >= batch_size:
                chunks.append(" ".join(current))
                current = []
        if current:
            chunks.append(" ".join(current))
        return [chunk for chunk in chunks if chunk]

    def process_text(
        self,
        text: str,
        config: ProcessingConfig,
        on_progress: Optional[Callable[[ProcessingProgress], None]] = None,
    ) -> ProcessingSummary:
        chunks = self.split_into_chunks(text, config.batch_size)
        total_chunks = len(chunks)
        processed: List[str] = []
        applied_steps: List[str] = []
        logs: List[str] = []

        total_input_tokens = 0
        total_output_tokens = 0
        total_cost = 0.0

        durations_total = 0.0

        for idx, chunk in enumerate(chunks):
            start_time = time.perf_counter()
            retry_count = 0
            success = False
            processed_chunk = chunk

            while retry_count < 2 and not success:
                try:
                    options = ProcessOptions(
                        text=chunk,
                        cleaning_options=config.cleaning_options,
                        speaker_config=config.speaker_config,
                        model_source=config.model_source,
                        model_name=config.model_name,
                        ollama_model_name=config.ollama_model_name,
                        temperature=config.temperature,
                        custom_instructions=config.custom_instructions,
                        single_pass=config.single_pass,
                        llm_cleaning_disabled=config.llm_cleaning_disabled,
                        extended_examples=config.extended_examples,
                    )
                    result = self.service.process_chunk(options)
                    processed_chunk = result.text
                    total_input_tokens += result.input_tokens
                    total_output_tokens += result.output_tokens
                    total_cost += result.input_cost + result.output_cost
                    applied_steps.extend(result.applied_steps)
                    success = self.service.validate_output(chunk, processed_chunk)
                except Exception as exc:  # noqa: BLE001 - surface any errors
                    retry_count += 1
                    logs.append(f"Chunk {idx + 1}: {type(exc).__name__}: {exc}")
                    if retry_count >= 2:
                        processed_chunk = chunk
                        logs.append(f"Chunk {idx + 1}: falling back to original text after errors.")
                        break
                    continue

                if not success:
                    retry_count += 1
                    time.sleep(1)

            processed.append(processed_chunk)
            elapsed = (time.perf_counter() - start_time) * 1000
            durations_total += elapsed
            avg_chunk_ms = durations_total / (idx + 1)
            remaining = total_chunks - (idx + 1)
            eta_ms = max(0, int(avg_chunk_ms * remaining))

            progress = ProcessingProgress(
                chunk_index=idx,
                processed_text=processed_chunk,
                status="success" if success else "failed",
                retry_count=retry_count,
                last_chunk_ms=int(elapsed),
                avg_chunk_ms=avg_chunk_ms,
                eta_ms=eta_ms,
                input_tokens=total_input_tokens,
                output_tokens=total_output_tokens,
                total_input_tokens=total_input_tokens,
                total_output_tokens=total_output_tokens,
                total_cost=total_cost,
            )
            if on_progress:
                on_progress(progress)

        final_text = "\n\n".join(processed)
        return ProcessingSummary(
            text=final_text,
            total_chunks=total_chunks,
            total_input_tokens=total_input_tokens,
            total_output_tokens=total_output_tokens,
            total_cost=total_cost,
            applied_cleaning_steps=applied_steps,
            logs=logs,
        )

    def deterministic_clean(self, text: str, options: CleaningOptions) -> ProcessingSummary:
        result = apply_deterministic_cleaning(text, options)
        return ProcessingSummary(
            text=result.text,
            total_chunks=1,
            total_input_tokens=0,
            total_output_tokens=0,
            total_cost=0.0,
            applied_cleaning_steps=result.applied,
            logs=[],
        )
